Deserialize

Ambrevar opened the pull request:
Fixes #16. 

As discussed, this fixes struct deserialization when they have a non-default constructor on SBCL, CCL, CLisp and ECL.
For other compilers, it raises an error.

Better: I've introduced 2 generic functions which allow for customizing the deserialization of structs and classes according to their symbol.

So if I want to specialize the deserialization of `quri.uri:uri`, I can specialize this way:

```lisp
(defmethod deserialize-struct ((struct-symbol (eql 'quri.uri:uri)) slots deserialized-objects)
  ...)
```

Thoughts?

Ambrevar committed:
Fix struct deserialization when non-default constructor is used.

Ambrevar committed:
Allow customizing deserialization with deserialize-class and deserialize-struct methods.

svetlyak40wt commented on src/serialization/sexp.lisp:
```diff
@@ -241,3 +232,32 @@
+                     #+(or sbcl ccl ecl clisp)
+                     (make-instance struct-symbol)
+                     #-(or sbcl ccl ecl clisp) ; ABCL and LispWorks don't support make-instance for structs.
+                     (error "Does not know how to deserialize non-default constructor"))))
```
Lets improve this error message like this:

```lisp
(error "Do not know how to deserialize struct ~S with non-default constructor"
           struct-symbol)
```

Ambrevar committed:
Improve deserialize-struct error message.

Ambrevar commented:
Done.

Ambrevar commented:
Don't merge!

Found a bug in Nyxt which breaks object reference when deserializing.

I'll investigate.

Ambrevar committed:
Initialize slots after updating the deserialized-objects id to the new object.

If we don't respect this order, it would fail to deserialize references in
object slots.

Ambrevar commented:
Fixed in 
df1da344ddf1b188c3ffc5c54808d94266356f79.

svetlyak40wt commented:
Ok, thank you!

Let's merge it now!

Ambrevar commented:
Thanks!
